### DISCUSSION TITLE
How can I get the final SQL executed by Drift?

### DISCUSSION THREAD

**Discussion by wvv8oo** [simolus3/drift#2506] [Q&A] [answered]
I want to get the actual SQL executed by Drift for debugging purposes, but I couldn't find this interface in the documentation?

**Comment by simolus3** [comment]
You can enable `logStatements: true` on all `NativeDatabase` constructors, which will make drift `print()` the SQL statement and variables before executing it.

**Reply by wvv8oo** [reply]
Thank you for your answer, but "enable logStatements: true" can only print the SQL statements during debugging, and cannot obtain the final generated SQL through code. For example, when a user encounters an exception or other scenarios, I want to report the SQL statements to the server for analysis, but I cannot implement this function. @simolus3

**Reply by simolus3** [reply, marked as answer]
You could implement your own `QueryExecutor` that takes an existing implementation from drift and augments it with logging/your custom error handling capabilities.

```dart
import 'package:drift/drift.dart';

class LoggingExecutor extends QueryExecutor {
  final QueryExecutor _inner;

  LoggingExecutor(this._inner);

  static QueryExecutor wrap(QueryExecutor inner) {
    if (inner is DatabaseConnection) {
      return inner.withExecutor(LoggingExecutor(inner.executor));
    } else {
      return LoggingExecutor(inner);
    }
  }

  Future<T> _log<T>(String description, Future<T> Function() inner) async {
    try {
      print('starting with $description');
      await inner();
    } catch (e) {
      // todo: Log again
      rethrow;
    }
  }

  @override
  TransactionExecutor beginTransaction() {
    return _LoggingTransactionExecutor(_inner.beginTransaction());
  }

  @override
  SqlDialect get dialect => _inner.dialect;

  @override
  Future<bool> ensureOpen(QueryExecutorUser user) {
    return _inner.ensureOpen(user);
  }

  @override
  Future<void> runBatched(BatchedStatements statements) {
    return _log(
        'Batched statements $statements', () => _inner.runBatched(statements));
  }

  @override
  Future<void> runCustom(String statement, [List<Object?>? args]) {
    return _log('Running $statement with $args',
        () => _inner.runCustom(statement, args));
  }

  // You can implement runDelete, runInsert, runSelect and runUpdate in a
  // similar fashion.
}

class _LoggingTransactionExecutor extends LoggingExecutor
    implements TransactionExecutor {
  final TransactionExecutor _innerTransaction;

  _LoggingTransactionExecutor(TransactionExecutor inner)
      : _innerTransaction = inner,
        super(inner);

  @override
  Future<void> rollback() {
    return _log('Rollback', () => _innerTransaction.rollback());
  }

  @override
  Future<void> send() {
    return _log('Commit', () => _innerTransaction.send());
  }

  @override
  bool get supportsNestedTransactions =>
      _innerTransaction.supportsNestedTransactions;
}
```

Then you can use `LoggingExecutor.wrap()` of whatever you've been using before (either the whole `LazyDatabase` or the `NativeDatabase.createInBackground`).

**Reply by wvv8oo** [reply]
Thank you very much for your reply. This should solve the problem, however, I suggest that this capability be provided as I believe it is a relatively common one.

**Reply by simolus3** [reply]
Yeah, it does come up fairly often. I've tried to improve the situation a bit in 29379e434a8d10ab11f13703c69ef39cce5157ed - it would allow you to extend or implement a `QueryInterceptor` class to add the functionality. Essentially, the only difference is that you don't need to implement two classes and have it all in one place.

Is that what you're looking for? It still requires a bunch of work to setup, but I don't see how something on this level could be provided by drift directly - after all, only you know the logging/error handling setup to use in your app.

**Reply by jointhejourney** [reply]
Hey @simolus3 +1 on that work you committed, I'd also be super interested in using something like that. Thank you for the great work!
